correct device pm control return value for unimplemented handlers

pabigot opened the pull request:
`device_pm_control_nop` indicates success for all operations without doing anything.  So does the weak system timer pm control handler.  Make them return an error so callers are not deceived into thinking the unmodified output parameter to `device_get_power_state()` actually represents something meaningful.

pabigot committed:
device_pm: correct nop documented behavior

The device_pm_control_nop function is documented to always return zero
regardless of operation.  However, when device_get_power_state() is
invoked with this control function it returns success leaving the
output parameter state unmodified, which may not be a valid device
state.

Document and implement that the nop control function returns -ENOTSUP
always.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
kernel: sys_clock: update weak pm control function

The weak implementation returns 0 for all operations without doing
anything, which incorrectly suggests that an operation like
device_get_power_state() returned an accurate description of the
system clock power state.  Return -ENOTSUP instead.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
FWIW this is so egregious a violation of expected behavior I think it's worth considering for 2.3.0 even if it is an "api change" for all the devices that don't support device power management.

carlescufi commented:
> FWIW this is so egregious a violation of expected behavior I think it's worth considering for 2.3.0 even if it is an "api change" for all the devices that don't support device power management.

I agree. It's technically an API change but this was clearly wrong and highly unlikely that anyone relied on the old error code, so I think we can safely include this for 2.3.0.

pabigot committed:
power: device: adapt to new behavior for devices with nop pm ctrl

For backwards compatibility ignore not-supported errors for devices
that don't support power management.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Apparently existing code including the power management infrastructure expects to see success.  Enabling device power management in conjunction with system power management with the original two patches prevented system power management from shutting anything down.  Fixed at least one failure point, but I'm withdrawing my suggestion that this should be fixed for the release.

pabigot committed:
tests: kernel: device: bypass device pm tests

The dummy driver never implemented device power management, so the
fact the not-implemented stub returned success was a false negative.
Verify the expected behavior now, leaving the test code in place for
when somebody provides a non-trivial PM control function.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented on tests/kernel/device/src/main.c:
```diff
@@ -208,6 +208,11 @@ void test_dummy_device_pm(void)
 	/* Set device state to DEVICE_PM_ACTIVE_STATE */
 	ret = device_set_power_state(dev, DEVICE_PM_ACTIVE_STATE, NULL, NULL);
+	if (ret == -ENOTSUP) {
+		zassert_true((ret == -ENOTSUP),
```
Yeah, I just wanted to leave some sort of assert in there, otherwise the test wouldn't check anything.
